### PR TITLE
merkleRoot() uses different compression for odd nodes and bottom nodes

### LINKED ISSUE #5: The Merkle root construction is somewhat unsafe
The Merkle root construction is somewhat unsafe, in the sense that calling the `merkleRoot` function with for example `[3,2,1]` gives the same result as calling with `[3,2,1,0]`.

So it can be only used for hashing fixed size arrays. 

Ideally we should change the construction so that there are no (practical) collisions between differently sized inputs. Some possible options:
- somehow include the length itself in the final root (for example adding a final `realRoot = hash(almostRoot,length)` step)
- pre-hash every element, and use 0 as _hashes_ as opposite to _values_ to pad (downside: makes it twice as slow)

If this is already solved in Codex, we should probably mirror that solution (but definitely synchronize the two)

Less ideally we should document this behaviour and/or enforce power-of-two-sized inputs.

### PR DESCRIPTION
**Pull request by markspanbroek**
Adds extra data to the compress function that indicates whether a node in the merkle tree is on the bottom layer and whether the node has an odd number of children.

Implements solution 3 from https://github.com/codex-storage/codex-research/pull/189 against merkle tree attacks.

~~Depends on #10.~~

Fixes #5

### CODE REVIEW

File: poseidon2/merkle.nim [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import ./io
+import ./compress
+
+const KeyNone = F.fromHex("0x0")
+const KeyBottomLayer = F.fromHex("0x1")
+const KeyOdd = F.fromHex("0x2")
+const KeyOddAndBottomLayer = F.fromhex("0x3")
+
+func merkleRoot(xs: openArray[F], isBottomLayer: static bool) : F =
+  let a = low(xs)
```
Comment (bkomuves) [inline poseidon2/merkle.nim:14]: `merkleRoot` appears to get into an infinite loop when the input is empty. Suggestion: handle that explicitly and throw an exception that the empty input is invalid

**Review by bkomuves — COMMENTED** [review]
Looks good to me in general.

However, I would like to have tests which test concrete input-output pairs for: Merkle tree of sequences of both field elements and bytes

I added "test vector generation" here: https://github.com/codex-storage/codex-storage-proofs-circuits/tree/master/reference/nim

Currently this matches the Haskell implementation in the same repo.

File: poseidon2/merkle.nim (hunk 1 above)
Comment (markspanbroek) [reply poseidon2/merkle.nim:14]: Good catch, thanks! Fixed in 7d02d4f

**Comment by markspanbroek** [comment]
> However, I would like to have tests which test concrete input-output pairs for: Merkle tree of sequences of both field elements and bytes

Added in 7d0ad4f

**Pull request merged**